File: src/main/java/com/karmanno/plugins/GitUtils.java
package com.karmanno.plugins;

import org.gradle.api.Project;
import org.gradle.internal.impldep.org.eclipse.jgit.api.Git;
import org.gradle.internal.impldep.org.eclipse.jgit.internal.storage.file.FileRepository;

import java.io.File;
import java.io.IOException;

public class GitUtils {
    public static Git gitRepo(Project project) {
        try {
            File gitDir = getRootGitDir(project.getProjectDir());
            return Git.wrap(new FileRepository(gitDir));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static File getRootGitDir(File currentRoot) {
        File gitDir = scanForRootGitDir(currentRoot);
        if (!gitDir.exists()) {
            throw new IllegalArgumentException("Cannot find '.git' directory");
        }
        return gitDir;
    }

    private static File scanForRootGitDir(File currentRoot) {
        File gitDir = new File(currentRoot, ".git");
        if (gitDir.exists()) {
            return gitDir;
        }
        if (currentRoot.getParentFile() == null) {
            return gitDir;
        }
        return scanForRootGitDir(currentRoot.getParentFile());
    }
}

File: src/main/kotlin/com/karmanno/plugins/GitUtils.kt
package com.karmanno.plugins

import org.eclipse.jgit.api.Git
import org.eclipse.jgit.internal.storage.file.FileRepository
import org.gradle.api.Project
import java.io.File

class GitUtils {
    companion object {
        fun repo(project: Project): Git = try {
            Git.wrap(FileRepository(gitDirectory(project.projectDir)))
        } catch (e: Exception) {
            project.logger.error("Couldn't acquire git directory for dir: ${project.projectDir}")
            throw RuntimeException(e)
        }

        private fun gitDirectory(directory: File): File {
            val gitDir = scanForGitRoot(directory)
            if (!gitDir.exists()) {
                throw IllegalArgumentException("Cannot find '.git' directory")
            }
            return gitDir
        }

        private fun scanForGitRoot(directory: File): File {
            val gitDir = File(directory, ".git")
            if (gitDir.exists()) {
                return gitDir
            }
            if (directory.parentFile == null) {
                return gitDir
            }
            return scanForGitRoot(directory.parentFile)
        }
    }
}